zerver/lib/actions.py

```diff
@@ -694,14 +694,12 @@ def do_create_user(
 
 
 def do_activate_user(user_profile: UserProfile, *, acting_user: Optional[UserProfile]) -> None:
-    user_profile.is_active = True
+    change_user_is_active(user_profile, True)
     user_profile.is_mirror_dummy = False
     user_profile.set_unusable_password()
     user_profile.date_joined = timezone_now()
     user_profile.tos_version = settings.TOS_VERSION
-    user_profile.save(
-        update_fields=["is_active", "date_joined", "password", "is_mirror_dummy", "tos_version"]
-    )
+    user_profile.save(update_fields=["date_joined", "password", "is_mirror_dummy", "tos_version"])
```

Comment:
For the transaction structure here, we really want the `with transaction.atomic` around the whole block where we're changing UserProfile (and RealmAuditLog as well) -- we don't want it to be possible to have `is_mirror_dummy=True`  for an active user.

(Similar issue for `do_deactivate_user`, I think).

Is this a case where nesting `transaction.atomic()` is OK?

Reply:
I guess one possible option is to not use `change_user_profile` for this code path and exclude it from semgrep with comments.

Reply:
Hmm, what's the main argument against nesting? As long as we use `savepoint=False` to avoid overhead, I don't see why nesting would be a big issue? I pushed a commit to run the user activation/reactivation/deactivation queries atomically.

Reply:
Maybe it's actually fine in this context.